rust/kernel/str.rs

```diff
@@ -0,0 +1,243 @@
+
+use crate::bindings;
+use crate::c_types;
+
+/// Byte string without UTF-8 validity guarantee.
+///
+/// `BStr` is simply an alias to `[u8]`, but has a more evident semantical meaning.
+pub type BStr = [u8];
+
+/// Creates a new [`BStr`] from a string literal.
+///
+/// `b_str!` converts the supplied string literal to byte string, so non-ASCII
+/// characters can be included.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// const MY_BSTR: &'static BStr = b_str!("My awesome BStr!");
+/// ```
+#[macro_export]
+macro_rules! b_str {
+    ($str:literal) => {{
+        const S: &'static str = $str;
+        const C: &'static $crate::str::BStr = S.as_bytes();
+        C
+    }};
+}
+
+/// Possible errors when using conversion functions in [`CStr`].
+#[derive(Debug, Clone, Copy)]
+pub enum CStrConvertError {
+    /// Supplied bytes contain an interior `NUL`.
+    InteriorNul,
+
+    /// Supplied bytes are not terminated by `NUL`.
+    NotNulTerminated,
+}
+
+impl From<CStrConvertError> for crate::Error {
+    #[inline]
+    fn from(_: CStrConvertError) -> crate::Error {
+        crate::Error::EINVAL
+    }
+}
+
+/// A string that is guaranteed to have exactly one `NUL` byte, which is at the
+/// end.
+///
+/// Used for interoperability with kernel APIs that take C strings.
+#[repr(transparent)]
+pub struct CStr([u8]);
+
+impl CStr {
+    /// Returns the length of this string excluding `NUL`.
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.0.len() - 1
+    }
+
+    /// Returns the length of this string with `NUL`.
+    #[inline]
+    pub const fn len_with_nul(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if the string only includes `NUL`.
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Wraps a raw C string pointer.
+    ///
+    /// # Safety
+    ///
+    /// `ptr` must be a valid pointer to a `NUL`-terminated C string, and it must
+    /// last at least `'a`. When `CStr` is alive, the memory pointed by `ptr`
+    /// must not be mutated.
+    #[inline]
+    pub unsafe fn from_char_ptr<'a>(ptr: *const c_types::c_char) -> &'a Self {
+        let len = bindings::strlen(ptr) + 1;
+        Self::from_bytes_with_nul_unchecked(core::slice::from_raw_parts(ptr as _, len as _))
+    }
+
+    /// Creates a [`CStr`] from a `[u8]`.
+    ///
+    /// The provided slice must be `NUL`-terminated, does not contain any
+    /// interior `NUL` bytes.
+    pub const fn from_bytes_with_nul(bytes: &[u8]) -> Result<&Self, CStrConvertError> {
+        if bytes.is_empty() {
+            return Err(CStrConvertError::NotNulTerminated);
+        }
+        if bytes[bytes.len() - 1] != 0 {
+            return Err(CStrConvertError::NotNulTerminated);
+        }
+        let mut i = 0;
+        while i < bytes.len() - 1 {
+            if bytes[i] == 0 {
+                return Err(CStrConvertError::InteriorNul);
+            }
+            i += 1;
+        }
```

Comment:
This could be
```rust
if bytes[..bytes.len() - 1].iter().any(|x| x == 0) {
    return Err(CStrConvertError::InteriorNul);
}
```
which I would ensure that there are no excessive bounds checks for `bytes[i]`.
(The compiler can _probably_ elide these here but I would generally rather not rely on it.)

Alternatively it could also be
```rust
for x in &bytes[..bytes.len() - 1] {
    if x == 0 {
        return Err(CStrConvertError::InteriorNul);
    }
}
```

Reply:
Slicing could not be used in const fn.

Reply:
Originally this is written as `bytes[..bytes.len() - 1].contains(&0)`, but following suggestion from @wedsonaf I changed it to a while loop as it currently is (for loops also couldn't be used in const fn).

Reply:
Ah, I missed that this has to be `const`, sorry.
I guess we'll just have to hope/check that the bounds checks are properly elided. (Or provide separate const/runtime versions but that doesn't seem very nice)

Reply:
> Originally this is written as bytes[..bytes.len() - 1].contains(&0)

Yeah that would have been even better, it's quite unfortunate that const support in rust is still so immature.

Reply:
I looked at the assembly and it seems that LLVM couldn't infer that `i < len - 1` implies `i < len`. This boils down to the fact that `len - 1` is not UB even when `len == 0`, so there isn't sufficient info to optimize out the boundary check. It's unfortunate that LLVM couldn't take advantage of the knowledge that `len > 0`.

However somehow `i + 1 < len` would allow LLVM to optimize away the bound check. Perhaps it's more common to have comparison against len so LLVM is better at doing it.

Reply:
That is indeed weird. I guess if `i + 1 < len` works we can just go with that and add a comment why it is written this way.

Alternatively we could of course also use `get_unchecked` to ensure there are no bounds checks but I don't know if that is actually worth the extra unsafetly here since it is technically possible without it.

Reply:
We could also maybe use a `NonZeroUsize` for the len which I would _hope_ LLVM can optimize properly.
We could probably even make this completely safe without losing perf by replacing the initial `is_empty()` check by with a `NonZeroUsize::new(bytes.len())` call.

Reply:
I don't think `get_unchecked` is worth it, and it is not const fn either. TBH I don't think this function'll be used in hot path; usually `CStr` is constructed either statically or from `char*` with `from_char_ptr`. So even bounds checks are left, it wouldn't be a big deal. It's too premature to optimize it. I would prefer simplicity to a few unnecessary bounds checks.

`NonZeroUsize` is usually for memory optimisations. LLVM already knows that length is not 0 after the is_empty check, it just simply couldn't use it to infer `i < len`, so replacing it with `NonZeroUsize` wouldn't help either.

Reply:
I agree that this is probably premature, I guess my usual use cases for CStr in regular Rust code may have made me overestimate the importance of this quite a bit.